protocol/l1-upgrades.md

```diff
@@ -0,0 +1,305 @@
+way of example, we take an initializable version of `SimpleStorage`, and refactor it into the set of
+contracts which would be required to specify a hypothetical Magma release, which adds a new variable
+to the contract.
+
+```solidity
+contract SimpleStorage is Initializable {
+  uint public foo;
+  bytes32 public bar; // new variable added for the Magma release.
+
+
+  /// @notice initialize is called whether for a fresh system or when upgrading an existing system.
+  function initialize(uint _foo, bytes32 bar) public initializer {
+    // chains being upgraded need to read the current value of foo, then write it again.
+    foo = _foo;
+    bar = _bar;
+  }
+}
+```
+
+This `SimpleStorage` will become:
+
+```solidity
+abstract contract SimpleStorageLayout {
+  uint public foo;
+  bytes32 public bar;
+}
+
+contract SimpleStoragePopulator is SimpleStorageLayout {
+  function initialize() public {
+    foo = 1;
+    bar = 2_000
+  }
+
+  function upgrade() public {
+    bar = 2_000;
+  }
+}
+
+contract SimpleStorage is SimpleStorageLayout {
+  // Whatever runtime logic is needed
+}
+```
+
+With this architecture, the OPCM contract (which will be a non-proxied singleton per release) would
+then either:
+
+1. Deploy a new contract by:
+   1. Deploying a new proxy
+   1. setting its implementation to the `SimpleStoragePopulator`
+   1. calling `initialize()`
+   1. setting its implementation to `SimpleStorage`
+1. Upgrade a contract by:
+   1. setting its implementation to the `SimpleStoragePopulator`
+   1. calling `upgrade()`
+   1. setting its implementation to `SimpleStorage`
+
+This workflow has the benefit of removing the `initializer()` functions, which:
+
+- create a risk of reinit attacks
+- bloat the runtime bytecode size
+- are extremely verbose
+- require finesse to avoid modifying or resetting values which are not meant to be changed.
+
+## A single `SuperchainProxyAdmin`
+
+The existing L1 `ProxyAdmin` (of which there is currently one per OP Chain), will be replaced with a
+single L1 `SuperchainProxyAdmin`. The `SuperchainProxyAdmin`, in contrast with the L1 `ProxyAdmin`,
+will have minimal storage, and will not internally track the type of Proxy which each contract is.
+
+This contract will be owned by the [Upgrade
+Controller](https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/stage-1.md#L30), and
+will be a simple pass through contract with a minimal interface.
+
+```
+/// @notice forwards calldata to the specified address.
+///   Reverts unless called by owner.
+function forward(address, bytes) external;
+
+/// @notice forwards calldata to the specified addresses.
+///   Reverts unless called by owner.
+function forward(address[], bytes[]) external;
+```
+
+This architecture makes it possible to manage upgrade authorization for the whole superchain through
+a single storage value in a single contract, rather than having a separate contract per OP Chain.
```

Comment:
Referring to the `owner` in the SuperchainProxyAdmin?

Another benefit is we can change the single proxy admin in the future to add new capabilities, such as a timelock to delay transactions

Reply:
Yes, and agree!